File: src/editor.ts
import { rm } from 'node:fs/promises'
import { tmpdir } from 'node:os'

// Open a file in $EDITOR and wait for it to close. `sh -c` so a multi-word
// $EDITOR (e.g. "code --wait") works; "$1" handles spaces in the path. $EDITOR
// must block until close (nano/vim do; GUI editors need a wait flag).
export async function openEditor(path: string): Promise<void> {
  const editor =
    process.env['AGENT_WORK_EDITOR'] || process.env['EDITOR'] || process.env['VISUAL'] || 'vi'
  const proc = Bun.spawn(['sh', '-c', `${editor} "$1"`, 'sh', path], {
    stdin: 'inherit',
    stdout: 'inherit',
    stderr: 'inherit',
  })
  const code = await proc.exited
  if (code !== 0) {
    throw new Error(`editor exited ${code}: ${editor}`)
  }
}

// Compose text in $EDITOR via a temp file, pre-filled with `seed`. Returns the
// trimmed contents ('' if left empty). Used for composing a task intent and for
// multi-line sharpen replies.
export async function composeInEditor(seed = ''): Promise<string> {
  const tmp = `${tmpdir()}/factory-edit-${Date.now()}.md`
  await Bun.write(tmp, seed ? `${seed}\n` : '')
  await openEditor(tmp)
  const text = (await Bun.file(tmp).text()).trim()
  await rm(tmp, { force: true })
  return text
}

File: src/editor.rs
use std::fs;
use std::io;
use std::path::Path;
use std::process::{Command, Stdio};

pub fn open_editor(path: impl AsRef<Path>) -> io::Result<()> {
    let editor = std::env::var("AGENT_WORK_EDITOR")
        .or_else(|_| std::env::var("EDITOR"))
        .or_else(|_| std::env::var("VISUAL"))
        .unwrap_or_else(|_| "vi".to_string());
    let status = Command::new("sh")
        .arg("-c")
        .arg(format!("{editor} \"$1\""))
        .arg("sh")
        .arg(path.as_ref())
        .stdin(Stdio::inherit())
        .stdout(Stdio::inherit())
        .stderr(Stdio::inherit())
        .status()?;
    if status.success() {
        Ok(())
    } else {
        Err(io::Error::other(format!(
            "editor exited {}: {editor}",
            status.code().unwrap_or(1)
        )))
    }
}

pub fn compose_in_editor(seed: &str) -> io::Result<String> {
    let tmp = std::env::temp_dir().join(format!(
        "factory-edit-{}.md",
        ::time::OffsetDateTime::now_utc().unix_timestamp_nanos()
    ));
    fs::write(
        &tmp,
        if seed.is_empty() {
            String::new()
        } else {
            format!("{seed}\n")
        },
    )?;
    let result = open_editor(&tmp).and_then(|()| fs::read_to_string(&tmp));
    fs::remove_file(&tmp).ok();
    result.map(|text| text.trim().to_string())
}
